cohorts/utils.py

```diff
@@ -0,0 +1,136 @@
+    # replace meaningful punctuation with text equivalents
+    # surround each with whitespace to enforce consistent use of _
+    punctuation_to_text = {
+        '<=': 'le',
+        '>=': 'ge',
+        '=<': 'le',
+        '=>': 'ge',
+        '<': 'lt',
+        '>': 'gt',
+        '#': 'num'
+    }
+    for punctuation, punctuation_text in punctuation_to_text.items():
+        new_col_name = new_col_name.replace(punctuation, punctuation_text)
+
+    # remove contents within ()
+    if not(keep_paren_contents):
+        new_col_name = re.sub('\([^)]*\)', '', new_col_name)
+
+    # replace remaining punctuation/whitespace with _
+    punct_pattern = '[\W_]+'
+    punct_replacement = '_'
+    new_col_name = re.sub(punct_pattern, punct_replacement, new_col_name)
+
+    # remove leading/trailing _ if it exists (if last char was punctuation)
+    new_col_name = new_col_name.strip("_")
+
+    # TODO: check for empty string
+
+    # return lower-case version of column name
+    return new_col_name.lower()
+
+
+def strip_column_names(cols, keep_paren_contents=True):
+    """
+    Utility script for renaming pandas columns to patsy-friendly names.
+
+    Revised names have been:
+        - stripped of all punctuation and whitespace (converted to text or `_`)
+        - converted to lower case
+
+    Takes a list of column names, returns a dict mapping
+    names to revised names.
+
+    If there are any concerns with the conversion, this will
+    print a warning & return original column names.
+
+    Parameters
+    ----------
+
+    cols (list): list of strings containing column names
+    keep_paren_contents (logical):
+        controls behavior of within-paren elements of text
+         - if True, (the default) all text within parens retained
+         - if False, text within parens will be removed from the field name
+
+    Returns
+    -------
+
+    dict mapping cols -> newcols
+
+    Example
+    -------
+
+    > df = {'one' : pd.Series([1., 2., 3.], index=['a', 'b', 'c']),
+      'two' : pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd']),
+      'PD L1 (value)': pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd']),
+      'PD L1 (>1)': pd.Series([0., 1., 1., 0.], index=['a', 'b', 'c', 'd']),
+      }
+    > df = pd.DataFrame(df)
+    > df = df.rename(columns = strip_column_names(df.columns))
+
+    ## observe, by comparison
+    > df2 = df.rename(columns = strip_column_names(df.columns,
+        keep_paren_contents=False))
+    """
+    logger = logging.getLogger()
+
+    # strip/replace punctuation
+    new_cols = [
+        _strip_column_name(col, keep_paren_contents=keep_paren_contents)
+        for col in cols]
+
+    if not(len(new_cols) == len(set(new_cols))):
+        logger.warning('Warning: strip_column_names (if run) would introduce duplicate names.'
+                       ' Reverting column names to the original.')
+        print('Warning: strip_column_names would introduce duplicate names. Please fix & try again.')
```

Comment:
Yes - that's why. I also ended up adding a `warning.warn` since the logger.warn message also didn't show up in the console -- appears this takes a separate setup which I was reading about here : https://docs.python.org/2.6/library/logging.html#configuring-logging

Definitely doesn't belong in this utils module, though should probably be addressed at some point .. for now I have found duplicating these messages helpful.